anaconda_anon_usage/patch.py

```diff
@@ -1,96 +1,42 @@
+# needed to deploy the additional anonmyous user data. It pulls
+# the token management functions themselves from the api module.
+
 import sys
-from os.path import dirname, exists, expanduser, join
-from uuid import uuid4
 
 from conda.auxlib.decorators import memoizedproperty
 from conda.base.context import Context, ParameterLoader, PrimitiveParameter, context
 
-from . import __version__
-
-DEBUG = bool(os.environ.get("ANACONDA_ANON_USAGE_DEBUG"))
-
-
-def _debug(s, *args, error=False):
-    if error or DEBUG:
-        print(s % args, file=sys.stderr)
-
-
-def get_random_token():
-    return base64.urlsafe_b64encode(uuid4().bytes).strip(b"=").decode("ascii")
-
-
-def get_saved_token(fpath, what):
-    client_token = ""
-    _debug("%s token path: %s", what.capitalize(), fpath)
-    if exists(fpath):
-        try:
-            # Use just the first line of the file, if it exists
-            client_token = "".join(open(fpath).read().splitlines()[:1])
-            _debug("Retrieved %s token: %s", what, client_token)
-        except Exception as exc:
-            _debug("Unexpected error reading: %s\n  %s", fpath, exc, error=True)
-    if len(client_token) < 22:
-        if len(client_token) > 0:
-            _debug("Generating longer token")
-        client_token = get_random_token()
-        try:
-            os.makedirs(dirname(fpath), exist_ok=True)
-            with open(fpath, "w") as fp:
-                fp.write(client_token)
-            _debug("Generated new token: %s", client_token)
-            _debug("%s token saved: %s", what.capitalize(), fpath)
-        except Exception as exc:
-            _debug("Unexpected error writing: %s\n  %s", fpath, exc, error=True)
-            client_token = ""
-    return client_token
-
-
-def get_client_token():
-    fpath = join(expanduser("~/.conda"), "aau_token")
-    return get_saved_token(fpath, "client")
-
-
-def get_environment_token():
-    try:
-        prefix = Context.checked_prefix or context.target_prefix
-        if prefix is None:
-            return None
-    except Exception as exc:
-        _debug("error retrieving prefix: %s", exc)
-        return None
-    fpath = join(prefix, "etc", "aau_token")
-    return get_saved_token(fpath, "environment")
-
-
-@functools.lru_cache(maxsize=None)
-def client_token_string():
-    parts = ["aau/" + __version__]
-    value = get_client_token()
-    if value:
-        parts.append("c/" + value)
-    parts.append("s/" + get_random_token())
-    value = get_environment_token()
-    if value:
-        parts.append("e/" + value)
-    result = " ".join(parts)
-    _debug("Full client token: %s", result)
-    return result
+from .api import _debug, token_string
 
 
 def _new_user_agent(ctx):
     result = ctx._old_user_agent
     if context.anaconda_anon_usage:
-        token = client_token_string()
-        if token:
-            result += " " + token
+        prefix = Context.checked_prefix or context.target_prefix or sys.prefix
```

Comment:
Is `sys.prefix` just another fallback?

Also, let's be extra careful about `checked_prefix` if, for some reason, the monkeypatch only half works, we can at least carry on here without erroring out

```suggestion
        prefix = getattr(Context, "checked_prefix", None) or getattr(
            context, "target_prefix", None
        ) or sys.prefix
```

Reply:
I would argue this is not necessary:

- target_prefix built-in. It always exists, even without the monkeypatch
- checked_prefix is set before `_new_user_agent` is slotted in. That means there is no way this code will be called if `checked_prefix` is not present.